src/java/org/apache/cassandra/index/sai/disk/v1/trie/TrieTermsDictionaryReader.java

```diff
@@ -0,0 +1,268 @@
+                                                  : Pair.create(collector.toByteComparable(), getCurrentPayload());
+            }
+
+            private long advanceNode()
+            {
+                long child;
+                int transitionByte;
+
+                go(stack.node);
+                while (true)
+                {
+                    int childIndex = stack.childIndex + 1;
+                    transitionByte = transitionByte(childIndex);
+
+                    if (transitionByte == Integer.MAX_VALUE)
+                    {
+                        // ascend
+                        stack = stack.prev;
+                        collector.pop();
+                        if (stack == null)
+                        {
+                            // exhausted whole trie
+                            return NOT_FOUND;
+                        }
+                        go(stack.node);
+                        continue;
+                    }
+
+                    child = transition(childIndex);
+                    stack.childIndex = childIndex;
+
+                    if (child != -1)
+                    {
+                        assert child >= 0 : String.format("Expected value >= 0 but got %d - %s", child, this);
+
+                        // descend
+                        go(child);
+
+                        stack = new IterationPosition(child, stack);
+                        collector.add(transitionByte);
+
+                        if (hasPayload())
+                            return child;
+                    }
+                }
+            }
+        };
+    }
+
+    public ByteComparable getMaxTerm()
+    {
+        TransitionBytesCollector collector = new ImmutableTransitionBytesCollector();
+        go(root);
+        while (true)
+        {
+            int lastIdx = transitionRange() - 1;
+            long lastChild = transition(lastIdx);
+            if (lastIdx < 0)
+            {
+                return collector.toByteComparable();
+            }
+            collector.add(transitionByte(lastIdx));
+            go(lastChild);
+        }
+    }
+
+    public ByteComparable getMinTerm()
+    {
+        TransitionBytesCollector collector = new ImmutableTransitionBytesCollector();
+        go(root);
+        while (true)
+        {
+            if (hasPayload())
+            {
+                return collector.toByteComparable();
+            }
+            collector.add(transitionByte(0));
+            go(transition(0));
+        }
+    }
+
+    private long getCurrentPayload()
+    {
+        return getPayload(buf, payloadPosition(), payloadFlags());
+    }
+
+    private long getPayload(ByteBuffer contents, int payloadPos, int bytes)
+    {
+        if (bytes == 0)
+        {
+            return NOT_FOUND;
+        }
+        return SizedInts.read(contents, payloadPos, bytes);
+    }
+
+    public static class ImmutableTransitionBytesCollector extends TransitionBytesCollector
```

Comment:
```suggestion
    public static class AppendOnlyTransitionBytesCollector extends TransitionBytesCollector
```
...or `AppendingTransitionBytesCollector`?

Reply:
This code has now been moved down to the `Walker` and this class removed and only the mutable version used. The immutable version of the collector made no real sense because the `pop` method couldn't be being called on the immutable version because it would throw an exception.